### ISSUE TITLE
[FR] - add Sketchup.require_relative & Sketchup.autoload

### ISSUE THREAD

**Issue by MSP-Greg** [SketchUp/api-issue-tracker#246]
Feature request/suggestion to add `Sketchup.require_relative` & `Sketchup.autoload`.

1. `Sketchup.require_relative` would simplify plugin code and also help with 'shared code' in known locations.

2. `Sketchup.autoload` would make it easier for people to set up their code to not load everything at once.  Maybe even a `Sketchup.autoload_relative` method.

For those not familiar with it, Ruby `autoload` takes two parameters, the first a constant symbol, the second a file path.  Essentially, the first time the constant is referenced, the file is loaded.  A common use might be a plugin that loads menu/toolbar items.  Maybe the toolbar has several buttons, and each uses a particular class/module.  Rather than load all the files when SU starts, `autoload` could be used to load the classes on demand, or when the user clicks a toolbar button.

Obviously, one could use a require, but `autoload` simplifies things.  In Ruby world, it also helps with CI, as when the app is done with autoload, there isn't a concern for loading/requiring files in the test framework...

**Comment by Eneroth3** [comment]
For require_relative you can use this method:

```ruby
# Load file relative to callers location. Unlike Kernel.require_relative this
# supports SketchUp's encrypted extension files.
#
# @param path [String]
#
# @return [Boolean] True when file is loaded. False if already loaded.
def self.require_relative(path)
  base = File.dirname(caller_locations(1, 1).first.absolute_path)
  path = File.join(base, path)

  Sketchup.require(path)
end
```

**Comment by MSP-Greg** [comment]
@Eneroth3

Thanks.  I'm aware there are ways to add it.

In general, since plugin 'packages' have encrypted files, having SketchUp equivalents to all Ruby loading methods would be helpful...

**Comment by thomthom** [comment]
In regard to `autoload`: https://bugs.ruby-lang.org/issues/5653
Matz has warned against using it, and wanted to remove it. Even without that warning I'm also not sure we can replicate the functionality for RBE/RBS files - would depend if the logic is exposed to Ruby's C API.

**Comment by thomthom** [comment]
Just noticed Matz added a comment 3 month ago:

> OK, I withdraw the proposal. The autoload method will stay (for Ruby3.0).

**Comment by MSP-Greg** [comment]
@thomthom

Yes, I participated in that discussion.  You might guess that I feel it should be available in Ruby.

Most of the issues involve using `autoload` in threads, and that's not something most people are doing with SU.  Also, because of that, most apps/gems aren't using it within threads.  But, yes, used incorrectly, it will break.

As mentioned, I would really like to see the Kernel methods monkey patched, so people pulling in gem code would have one less thing to change in the code.

I suppose another option would be to have some kind of comment flag in rb files that tells the extension packager/encrypter to not encrypt a file...

**Comment by thomthom** [comment]
I was looking at `autoload` earlier - because I was timing the loading time of extensions. Some of my own extensions can have hundreds of files. Combine that with any number of extensions a user might have and this starts to eat on the loading time. Loading on demand would be good for SU startup times.

Seeing how Matz had previously warned about `autoload` I didn't pursue that any further. Mainly because I didn't want to start using something that looked to be taken out at some point.
As for SketchUp - the Ruby API only works for the main thread, and we have that ongoing thread issue on Windows. So people have to use a single thread anyway.

I'd have to dig into the Ruby source to get a better idea of whether it's possible to do something like `Sketchup.autoload` to support autoload of RBS/RBE files.

**Comment by MSP-Greg** [comment]
> Loading on demand would be good for SU startup times.

Agreed.  Web apps are one thing, as some people feel those should 'fully load' at start.  SU file loading is normally UI based, so another 100 or 200 ms isn't an issue.

One thing about 'load-on-demand', with Ruby, there are tables of ids for methods, symbols, etc.   Loading new code does force them to be rebuilt, but with newer Ruby versions (and today's cpus & memory), it really isn't an issue for UI apps...

Also, if everything loaded on demand, it would probably be easier to determine if a plugin corrupted the environment...

Mostly addressed `autoload`, any thoughts on `require_relative`?  There were issues long ago if it was used with paths that were symlinks, but those issues have been fixed for at least a couple of versions...

**Comment by thomthom** [comment]
I'd think `require_relative` should be fairly straight forward to do.

**Comment by MSP-Greg** [comment]
Thanks.  If you want to close, fine by me...
